### PR TITLE
Implement ToImmutable*Async extension methods

### LINKED ISSUE #1540: Add ToImmutable*Async extension methods to System.Linq.Async
I would like to propose the addition of new extension methods that allow collecting an `IAsyncEnumerable` into the immutable collections from `System.Collections.Immutable`. 

The overloads below are copied from `System.Collections.Immutable` with a few changes to make them async friendly.

<details>

* Add `Async` suffix
* Change return type to `ValueTask`
* Add `CancellationToken` parameter

</details>

```cs
public static ValueTask<ImmutableArray<TSource>> ToImmutableArrayAsync<TSource>(this IAsyncEnumerable<TSource> source, CancellationToken cancellationToken = default);

public static ValueTask<ImmutableDictionary<TKey, TValue>> ToImmutableDictionaryAsync<TSource, TKey, TValue>(this IAsyncEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TSource, TValue> elementSelector, IEqualityComparer<TKey>? keyComparer, IEqualityComparer<TValue>? valueComparer, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableDictionary<TKey, TValue>> ToImmutableDictionaryAsync<TSource, TKey, TValue>(this IAsyncEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TSource, TValue> elementSelector, IEqualityComparer<TKey>? keyComparer, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableDictionary<TKey, TSource>> ToImmutableDictionaryAsync<TSource, TKey>(this IAsyncEnumerable<TSource> source, Func<TSource, TKey> keySelector, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableDictionary<TKey, TSource>> ToImmutableDictionaryAsync<TSource, TKey>(this IAsyncEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? keyComparer, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableDictionary<TKey, TValue>> ToImmutableDictionaryAsync<TSource, TKey, TValue>(this IAsyncEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TSource, TValue> elementSelector, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableDictionary<TKey, TValue>> ToImmutableDictionaryAsync<TKey, TValue>(this IAsyncEnumerable<KeyValuePair<TKey, TValue>> source, IEqualityComparer<TKey>? keyComparer, IEqualityComparer<TValue>? valueComparer, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableDictionary<TKey, TValue>> ToImmutableDictionaryAsync<TKey, TValue>(this IAsyncEnumerable<KeyValuePair<TKey, TValue>> source, IEqualityComparer<TKey>? keyComparer, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableDictionary<TKey, TValue>> ToImmutableDictionaryAsync<TKey, TValue>(this IAsyncEnumerable<KeyValuePair<TKey, TValue>> source, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableHashSet<TSource>> ToImmutableHashSetAsync<TSource>(this IAsyncEnumerable<TSource> source, IEqualityComparer<TSource>? equalityComparer, CancellationToken cancellationToken = default);

public static ValueTask<ImmutableHashSet<TSource>> ToImmutableHashSetAsync<TSource>(this IAsyncEnumerable<TSource> source, CancellationToken cancellationToken = default);

public static ValueTask<ImmutableList<TSource>> ToImmutableListAsync<TSource>(this IAsyncEnumerable<TSource> source, CancellationToken cancellationToken = default);

public static ValueTask<ImmutableSortedDictionary<TKey, TValue>> ToImmutableSortedDictionaryAsync<TSource, TKey, TValue>(this IAsyncEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TSource, TValue> elementSelector, IComparer<TKey>? keyComparer, IEqualityComparer<TValue>? valueComparer, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableSortedDictionary<TKey, TValue>> ToImmutableSortedDictionaryAsync<TSource, TKey, TValue>(this IAsyncEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TSource, TValue> elementSelector, IComparer<TKey>? keyComparer, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableSortedDictionary<TKey, TValue>> ToImmutableSortedDictionaryAsync<TSource, TKey, TValue>(this IAsyncEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TSource, TValue> elementSelector, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableSortedDictionary<TKey, TValue>> ToImmutableSortedDictionaryAsync<TKey, TValue>(this IAsyncEnumerable<KeyValuePair<TKey, TValue>> source, IComparer<TKey>? keyComparer, IEqualityComparer<TValue>? valueComparer, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableSortedDictionary<TKey, TValue>> ToImmutableSortedDictionaryAsync<TKey, TValue>(this IAsyncEnumerable<KeyValuePair<TKey, TValue>> source, IComparer<TKey>? keyComparer, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableSortedDictionary<TKey, TValue>> ToImmutableSortedDictionaryAsync<TKey, TValue>(this IAsyncEnumerable<KeyValuePair<TKey, TValue>> source, CancellationToken cancellationToken = default) where TKey : notnull;

public static ValueTask<ImmutableSortedSet<TSource>> ToImmutableSortedSetAsync<TSource>(this IAsyncEnumerable<TSource> source, IComparer<TSource>? comparer, CancellationToken cancellationToken = default);

public static ValueTask<ImmutableSortedSet<TSource>> ToImmutableSortedSetAsync<TSource>(this IAsyncEnumerable<TSource> source, CancellationToken cancellationToken = default);
```

## Pros/Cons
* Pro: Feature parity with the extensions available in the .NET SDK.
* Con: Requires a new NuGet dependency for older target frameworks because they don't include `System.Collections.Immutable`.
* others?

I'd be open to creating a PR for this feature.

### PR DESCRIPTION
**Pull request by bash**
Resolves #1540.

I copied most of the XML documentation and tests of the non-immutable equivalents (with adjustments of course).

b667720 contains a bit of a refactoring, but I'm not sure if it's a big improvement.

Working with `AsyncOverloads.tt` was a bit unpleasant, especially since the generated code has diverged from the generated code (nullability and XML documentation) which meant I had to be careful of what to commit and what to discard.
Wouldn't a source generator be a better fit, because it knows about nullability and generic constraints?

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/reactive?pullRequest=1545) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/reactive?pullRequest=1545) before we can accept your contribution.<br/><br/>:x: bash [sign now](https://cla.dotnetfoundation.org/dotnet/reactive?pullRequest=1545)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/reactive?pullRequest=1545) it.</sub>

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/reactive?pullRequest=1545) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/reactive?pullRequest=1545) before we can accept your contribution.<br/><br/>:x: bash [sign now](https://cla.dotnetfoundation.org/dotnet/reactive?pullRequest=1545)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/reactive?pullRequest=1545) it.</sub>

File: Ix.NET/Source/System.Linq.Async/System/Linq/FunctionalHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT License.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Async.System.Linq
```
Comment (bartdesmet) [inline Ix.NET/Source/System.Linq.Async/System/Linq/FunctionalHelpers.cs:9]: Clean up the namespace to `System.Linq`. I think this is the by-product of Visual Studio trying to guess the namespace when creating a new `.cs` file in a folder while the `csproj` does not specify some default namespace.

**Review by bartdesmet — CHANGES_REQUESTED** [review]
LGTM modulo a cosmetic nit around a namespace.

File: Ix.NET/Source/System.Linq.Async/System/Linq/FunctionalHelpers.cs (hunk 1 above)
Comment (bash) [reply Ix.NET/Source/System.Linq.Async/System/Linq/FunctionalHelpers.cs]: Ah yes I totally missed this 😄. Addressed in 1707045

**Review by bartdesmet — APPROVED** [review]

**Comment by bartdesmet** [comment]
A source generator could be a viable option to replace the T4. We've had these T4 files for years but as you've observed sometimes manual changes to the .cs files made their way in and slipped through the cracks in code reviews, so we should either do a pass to refresh them, or replace them. In the latter case, we need to make we end up generating the exact same code of course.

**Comment by bartdesmet** [comment]
@bash I've merged the PR for the source generators (thanks!), so could you go ahead and rebase this PR (due to the delete/edit conflicts for the T4 files which are now obsolete)?

**Comment by bash** [comment]
@bartdesmet Hurray! I'm really happy how the source generators turned out. This PR is all rebased and ready for another review. 😃

File: Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToImmutableDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+// The .NET Foundation licenses this file to you under the MIT License.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using static System.Linq.FunctionalHelpers;
+
+namespace System.Linq
```
[the remark this answers is not in the document]
Comment (bash) [reply Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToImmutableDictionary.cs]: Yess I finally got around to address your comment 🥳  (and fix my source generator ^^)

Oh, I haven't thought about this! I applied your suggestion for the namespace change and class name change. I also moved the files to match the new namespace.

There's not much that I know about how to best divide up assemblies, but it seems ok to me too to depend on `System.Collections.Immutable` conditionally, if it's only for one target platform.

**Comment by Romfos** [comment]
one year later -_-

**Comment by stijnherreman** [comment]
> one year later -_-

Looks like there are no active maintainers currently :(

**Comment by akarnokd** [comment]
Does this have to live in dotnet/reactive? C# has extension methods so unlike other Rx incarnations, there is no loss of convenience having it in a 3rd party library.

**Comment by stijnherreman** [comment]
That would hurt discoverability (as pointed out in the PR review), unless it's considered acceptable for a third party library to place things in a `System` namespace?

On the other hand, I wonder if this and the existing mutable methods can be moved to dotnet/runtime instead, now that `IAsyncEnumerable<T>` is available out of the box.

**Comment by Romfos** [comment]
> Does this have to live in dotnet/reactive? C# has extension methods so unlike other Rx incarnations, there is no loss of convenience having it in a 3rd party library.

I think yes, .NET 6 have this method out of the box for linq

**Comment by bash** [comment]
I always assumed that the goal of System.Linq.Async is to provide everything for async enumerables that Linq provides for enumerables (Although technically the `ToImmutable*` extension methods are part of `System.Collections.Immutable`).

I'd be willing to move this to a separate project if this should not be part of System.Linq.Async and someone has a good name suggestion for the package :)

**Comment by bash** [comment]
I'm no longer interested in updating this PR should a maintainer against all odds choose to review it.
Anyone is free to adopt the PR though in that case :)

**Pull request merged**